cmd/snap/inhibit.go

```diff
@@ -34,143 +34,153 @@ import (
 	"github.com/snapcore/snapd/usersession/client"
 )
 
+var runinhibitWaitWhileInhibited = runinhibit.WaitWhileInhibited
+
 func waitWhileInhibited(snapName string) error {
-	hint, _, err := runinhibit.IsLocked(snapName)
-	if err != nil {
-		return err
-	}
-	if hint == runinhibit.HintNotInhibited {
-		return nil
+	flow := newInhibitionFlow(snapName)
+	notified := false
+	inhibited := func(hint runinhibit.Hint, inhibitInfo *runinhibit.InhibitInfo) (cont bool, err error) {
+		if !notified {
+			// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
+			// when it has finished; the hook starts with HintInhibitedGateRefresh lock
+			// and then either unlocks it or changes to HintInhibitedForRefresh (see
+			// gateAutoRefreshHookHandler in hooks.go).
+			if hint != runinhibit.HintInhibitedForRefresh {
+				return false, nil
+			}
+			// Start notification flow.
+			if err := flow.Notify(); err != nil {
+				return true, err
+			}
+			// Make sure we call notification flow only once.
+			notified = true
+		}
+		return false, nil
 	}
 
-	// wait for HintInhibitedForRefresh set by gate-auto-refresh hook handler
-	// when it has finished; the hook starts with HintInhibitedGateRefresh lock
-	// and then either unlocks it or changes to HintInhibitedForRefresh (see
-	// gateAutoRefreshHookHandler in hooks.go).
-	// waitInhibitUnlock will return also on HintNotInhibited.
-	notInhibited, err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh)
+	hintFlock, err := runinhibitWaitWhileInhibited(snapName, nil, inhibited, 500*time.Millisecond)
 	if err != nil {
+		// It is fine to return an error here without finishing the notification
+		// flow because we either failed because of it or before it, so it
+		// should not have started in the first place.
 		return err
 	}
-	if notInhibited {
-		return nil
-	}
 
-	if isGraphicalSession() {
-		notifiedDesktopIntegration, err := tryNotifyRefreshViaSnapDesktopIntegrationFlow(snapName)
-		if err != nil {
+	// XXX: closing as we don't need it for now, this lock will be used in a later iteration
+	hintFlock.Close()
+
+	if notified {
+		if err := flow.Finish(); err != nil {
 			return err
 		}
-		if notifiedDesktopIntegration {
-			return nil
-		}
-		return graphicalSessionFlow(snapName, hint)
 	}
-	// terminal and headless
-	return textFlow(snapName, hint)
+
+	return nil
+}
+
+type inhibitionFlow interface {
```

Comment:
I am not aware of instances were it might clash, but I agree that it can be a bit more descriptive. How about `NotifyStart/NotifyFinish` or `InhibitNotifyStart/InhibitNotifyFinish`? I am avoiding `InhibitFinish` to avoid the confusion with actually uninhibited an inhibited snap.

I am not sure about `RefreshInhibit` because a snap could be inhibited for a few reasons.

Reply:
I just mean that since interfaces are implicit, having shorter method names may accidentally clash at some point. In this case `Notify` is such a generic word that I would probably couple it with a noun of some sort to make it clearer what is being notified.

Reply:
I took a look around the codebase I found the pattern for actions always starts with a verb so I updated the interface to `{Start,Finish}InhibitionNotification`, I would love to know your suggestion.

Reply:
I think those are fine because they are descriptive and not entirely generic as `Notify` is.